packages/artifact/src/internal-http-manager.ts

```diff
@@ -0,0 +1,36 @@
+import {HttpClient} from '@actions/http-client/index'
+import {createHttpClient} from './internal-utils'
+
+/**
+ * Used for managing all http Connections during either upload or download in order to limit the number of tcp connections created
```

Comment:
Out of curiosity, why do we need multiple http connections?

Does Node not handle multiple requests on a single client well?

Reply:
Multiple connections are setup so that it's easier to reset/dispose of one if something goes wrong.

We use the `keep-alive` header to not close the connection and to increase responsiveness. If something goes wrong, the connection should be disposed of. See https://github.com/microsoft/typed-rest-client/blob/5ce8cd7ae336384bd2d1851f5e845d89885c3d21/lib/HttpClient.ts#L313 (same thing applies to the rest client we use)

Disposing of one client should not interfere with any other concurrent uploads going on with another client so that's why multiple connections are used.

Using a single connection for every concurrent upload was also a recommendation that I got from @ericsciple when I was fighting with too many connections being open and weird timeouts

Reply:
>Using a single connection for every concurrent upload was also a recommendation

Would it be worth going down that route? It would simplify a lot of this code

Reply:
That's what I have in my implementation. There is effectively an array of http clients and each concurrent upload gets it's own. 

Initialized here: https://github.com/actions/toolkit/blob/2bc6c35d66e24f5fce0af8201e13284ce6a8e206/packages/artifact/src/internal/upload-http-client.ts#L119

Reply:
Why not share a single http client among all concurrent uploads?